### ISSUE TITLE
First class support for matrix

### ISSUE THREAD

**Issue by erlend-sh** [gitterHQ/gitter#1684]
In light of Gitter [going fully open source and more service-agnostic](http://blog.gitter.im/2017/03/15/gitter-gitlab-acquisition/) (no longer exclusive to GitHub) in the near future, I'd like to put in a feature request for a first class "matrix-bridge" project, equivalent to the [irc-bridge](https://github.com/gitterHQ/irc-bridge) for the IRC protocol.

Chat software is at a critical point right now.

- [Freenode is under the PIA umbrella](https://freenode.net/news/pia-fn). While I ultimately think this is a huge win for Freenode (yay for sustainable open source!) it does mean Freenode is a little less *free* than before. I trust the Freenode devs & PIA to continue to shepherd the Freenode project & community responsibly, but at the same time I firmly believe that competitive alternatives are the best way for any project to "stay true to itself".

- Slack is a great tool (I use it every day for work), but it's [not good for open source](http://sircmpwn.github.io/2015/11/01/Please-stop-using-slack.html). No [really, it's just not](https://dave.cheney.net/2017/04/11/why-slack-is-inappropriate-for-open-source-communications).

With Gitter going open source, backed by a truly open source company, Gitter could easily become the spiritual successor of Freenode. More user friendly; just as open; sustainable and, lastly, *multi-protocol*.

I'm aware of #1076 and [matrix-appservice-gitter](https://github.com/matrix-org/matrix-appservice-gitter), but all of these solutions require a degree of setup that most projects can't be bothered with. By far the best way to make the Matrix protocol accessible would be to support it to the same degree that you already support IRC.

**Comment by MadLittleMods** [comment]
IRC is supported via the already fully available developer API, https://github.com/gitterHQ/irc-bridge, https://developer.gitter.im/docs/welcome

From your issue, it seems like you are need of #1076 to get first-class support. Probably would be better to create separate issues for any other necessary features.

**Comment by erlend-sh** [comment]
I'm confused. Yes, IRC is already fully supported. There's no equivalent of "irc-bridge" for Matrix, and that's what I wanted to discuss here.

While #1076 might be necessary for a hypothetical "matrix-bridge" to work, the out-of-the-box support for Matrix would still require "matrix-bridge" to be an official Gitter project, run as a service by Gitter, just as [the IRC server](https://irc.gitter.im/) is.

**Comment by ara4n** [comment]
I think the confusion here is the definition of 'first class' support.  If gitter spoke the Matrix federation API natively (i.e. first class support), then it could directly share conversations with the wider open Matrix network.  On the Matrix side, we've just written a golang federation library to help with this (https://github.com/matrix-org/gomatrixserverlib) and there are plans to write a similar one for JS in future.  Meanwhile we're fixing up the federation spec as part of this work to make it easier for people to write federation implementations, although this is relatively complicated - it's similar complexity to writing a database replication system (given that's what it is).

Meanwhile, there's second class support available today via matrix-appservice-gitter (and other bridges), which would indeed benefit from #1076.  However, these require the user to run a matrix homeserver (e.g. synapse) as well as the bridge (matrix-appservice-gitter) as well as the actual chat server (gitter itself), which is not exactly elegant, even though it's the best we have today.  Hence (I assume) the request for first class support from gitter instead :)

@MadLittleMods any chance of reopening this?

**Comment by ara4n** [comment]
thanks for reopening, although the renaming is a bit confusing: "matrix bridge" is the terminology used to describe second class bridging into Matrix (i.e. via bridge application).  A better description might be "Support for federation with Matrix" or "Implement the Matrix federation protocol for interoperability with the wider Matrix ecosystem".

**Comment by MadLittleMods** [comment]
@ara4n If we do want to add Matrix support, it is likely to follow the same route as we did for IRC with a [separate bridge](https://github.com/gitterHQ/irc-bridge), just like `irc.gitter.im` as @erlend-sh mentioned.

**Comment by ara4n** [comment]
It's important to understand that IRC & Matrix, whilst cosmetically slightly similar, are architecturally entirely different.  irc.gitter.im is bridge that lets IRC clients connect to gitter servers.  Nobody is proposing this for Matrix - it wouldn't be particularly useful to just be able to point a Matrix client like riot.im to the channels sitting on a given gitter server.

Instead, the request here is to actually federate gitter instances with the rest of Matrix - exposing the channels so that they can be interact with users/bots/services/bridges in the wider Matrix network.  For instance, this could provide seamless bridging between gitter and existing IRC networks, Slack channels, Telegram group chats, XMPP MUCs etc.  Meanwhile, gitter users would get to keep and enjoy all the gitter-specific features and UI/UX, but the wider world would at least be able to follow along via a client/app of their choice.

So the way to do this would be to run a matrix federation component of some kind in gitter, which exposed the conversations appropriately.  I'm deliberately shying away of the word bridge, given in practice Gitter & Matrix use the term to mean different things :)

**Comment by MadLittleMods** [comment]
I am not familiar with Matrix. I assume you are after [application service API](https://matrix.org/docs/spec/application_service/unstable.html), *"defines how to extend the functionality of Matrix with ‘integrations’ and bridge to other networks"*, being covered (I haven't read through it). Which pieces can't be covered by a service, `matrix.gitter.im`, and communicate with the existing Gitter developer API?

**Comment by ara4n** [comment]
TL;DR: I think @erlend-sh is asking for Gitter to speak the Matrix [federation API](http://matrix.org/docs/spec/server_server/unstable.html), not the application service API.

So: Matrix is effectively an open standard & open decentralised network for publishing & subscribing to arbitrary data in realtime.  You can think of it a bit like decentralised Firebase or Pusher/Pubnub etc.  One use of it is to exchange chat history.  Matrix exposes 3 different APIs:

1. the [Client-Server (CS) API](http://matrix.org/docs/spec/client_server/r0.2.0.html), which lets native Matrix clients talk to Matrix homeservers - e.g. a chat app like Riot.im talk to a homeserver such as Synapse.  This is conceptually similar to the normal IRC client protocol exposed by irc.gitter.im.  There's not a huge amount of point in exposing this API given you already have the Gitter webapp!

2. the [Application Service (AS) API](http://matrix.org/docs/spec/application_service/unstable.html), which lets people write bridges which connect existing services like Gitter/Slack/IRC/Telegram/etc with the rest of Matrix.  Architecturally these bridges are easy to write and exist today, and the topology is:  Matrix Server <=AS=> Bridge <==> Gitter.  In other words, the person running the bridge has to run both a gitter server, a bridge server (i.e. matrix-appservice-gitter), and a matrix homeserver.  This is doable using the current gitter developer API but a bit clunky - why should the user have to run two separate messaging servers?  It also suffers from limitations like #1076.  

3. the [Federation (Server-Server aka SS) API](http://matrix.org/docs/spec/server_server/unstable.html), which lets Matrix-capable servers directly synchronise history with one another.  It's more complicated than the others, as Matrix history is stored in a cryptographically signed Merkle DAG and is synchronised using eventual consistency semantics in a manner very similar to Git.  If gitter implemented it, it could directly interface with the rest of Matrix natively without users having to also run a separate matrix homeserver or application service.

In practice, you'd need to wait for someone (possibly us) to publish a JS library for performing matrix federation, and then someone would plug it into the gitter messaging server (coupled as tightly as is appropriate).  But then in exchange you'd magically have gitter instances not acting as mini silos, but participating in an entirely decentralised heterogenous wider network.

**Comment by erlend-sh** [comment]
I get that (3) would be ideal, but it's a much tougher sell. In (2), you say:

> This is doable using the current gitter developer API but a bit clunky - why should the user have to run two separate messaging servers?

Since Gitter is a centralised service, the only "user" that needs to run two separate servers is the Gitter service, like they do with IRC already. Right?

And it makes sense for Gitter to build that kind of infrastructure, because they're in many ways trying to be a new type of Freenode, but more accessible.

Honestly I don't know which API is the right fit for what I'm asking for. I can only speak to the use case that I'm after:

- If I make a new Gitter room, e.g. https://gitter.im/myproject/myproject then..
- Matrix users can also join this room at https://riot.im/app/#/myproject/#myproject:gitter.im or through any other Matrix client.

**Comment by ara4n** [comment]
Well, I was thinking ahead to a world where users are running their own opensource gitter server instances.

The use case you mention can be supported today through Matrix's Application-Service API however - Matrix.org already hosts a Gitter bridge (matrix-appservice-gitter) pointed at gitter.im.  Right now you have to explicitly bridge a Matrix room into its Gitter counterpart but it works already.  It suffers from #1076 meaning that the Matrix users are presented all as a generic to Gitter users, however I hope this will be fixed someday (even if I have to submit the PR against Gitter myself ;).  We're also adding in 'puppeting' support to matrix-appservice-gitter to let users log in with their Gitter credentials if they want to be presented as their real gitter user on the gitter side of the bridge.

There's a lot more detail on this at https://matrix.org/blog/2017/03/11/how-do-i-bridge-thee-let-me-count-the-ways/.

If gitter wanted to they could obviously host their own Matrix homeserver & matrix-appservice-gitter deployment, but I imagine they have enough work keeping their own service running without also managing federation with Matrix! :)

**Comment by erlend-sh** [comment]
> Right now you have to explicitly bridge a Matrix room into its Gitter counterpart but it works already.

Yes, but with a Gitter-hosted homeserver that moderately complicated setup could be reduced to a few button clicks, and that's a hugely improved user experience.

I imagine a "baked in" server might also mitigate the login issue: If I log in on Riot.im as just a guest, I can't join any Gitter channel. That's a pretty big discrepancy.

Maybe even that is waaay too much work for just an onboarding gain; I don't know what GitLab's grand plan for Gitter is ;)

In any case, #1076 would be a great advancement no matter what, so I sincerely hope that'll get some attention as soon as Gitter goes fully open source.

**Comment by ara4n** [comment]
So we already have 'portal' access to Gitter set up on the matrix-appservice-gitter hosted by Matrix.org - any user on (say) Riot can `/join #gitter_ipfs=2Fipfs:matrix.org` or similar to jump into any arbitrary gitter room.  There's a sensible UI too in the room directory; you just select gitter from the dropdown list of networks and enter a channel identifier and jump in.  The behaviour would be identical if Gitter themselves ran the bridge - the room aliases would just be `#ipfs=2Fipfs:gitter.im` instead (or perhaps #ipfs_ipfs:gitter.im - the / makes for a terrible separator character!)

We deliberately don't let guests join though, as they haven't gone through a captcha and we don't want abuse.  However, we're currently reworking guest access to improve this (and force them through a captcha anyway) - https://github.com/vector-im/riot-meta/issues/59 has the details.

To conclude: #1076 is the biggest thing getting in the way of a good UX.
